README.en

```diff
@@ -0,0 +1,357 @@
+web crawler and became an independent project. 
+
+
+Quick start
+=============
+
+#include <stdio.h>
+#include <vector>
+#include <pire/pire.h>
+
+Pire::Scanner CompileRegexp(const char* pattern)
+{
+	// Convert the template to UCS4
+	std::vector<Pire::wchar32> ucs4;
+	Pire::Encodings::Utf8().FromLocal(pattern, pattern + strlen(pattern), std::back_inserter(ucs4));
+	// или другая кодировка
+
+	return Pire::Lexer(ucs4.begin(), ucs4.end())
+		.AddFeature(Pire::Features::CaseInsensitive())	// to ignore case
+		.SetEncoding(Pire::Encodings::Utf8())		// set the input string encoding
+		.Parse() 					// parse the regular expression
+		.Surround()					// to disable the PCRE_ANCHORED logic
+		.Compile<Pire::Scanner>();			// compile the regular expression
+}
+
+bool Matches(const Pire::Scanner& scanner, const char* ptr, size_t len)
+{
+	return Pire::Runner(scanner)
+		.Begin()	// the start of the string
+		.Run(ptr, len)	// the string
+		.End();		// the end of the string
+		// implicitly cast to bool
+}
+
+int main()
+{
+	char re[] = "hello\\s+w.+d$";
+	char str[] = "Hello world";
+
+	Pire::Scanner sc = CompileRegexp(re);
+
+	bool res = Matches(sc, str, strlen(str));
+
+	printf("String \"%s\" %s \"%s\"\n", str, (res ? "matches" : "doesn't match"), re);
+		
+	return 0;
+}
+
+
+Running `MATCH`
+==================
+
+The compiled regular expression is stored in the scanner which is the state machine
+in the form optimized for moving across the strings. There are several scanners with
+approximately the same operation in Pire (for example, Scanner, SlowScanner и SimpleScanner).
+
+Each scanner has the State type that stores the state, and methods Initialize(State&)
+to return the initial machine state, Next(State&, char) to take the machine to the next state,
+Final(const State&) to report whether the state is valid. To check whether the string matches
+the pattern, it is required to run Initialize(), then Next() for each letter, then Final()
+to get the result.
+
+There is the Pire::Run(const Scanner&, Scanner::State&, const char* begin, const char* end)
+method that runs optimized Scanner::Next() for each character from the range.
+
+The pattern may include characters of the string start and end (^ and $ respectively).
+These characters are replaced with BeginMark and EndMark at compilation.
+
+To check the string with two scanners, you can use the Pire::Run(const Scanner1&, 
+const Scanner2&, Scanner1::State&, Scanner2::State&, const char* begin, const char* end)
+function. This function checks the string faster than two consecutive calls of
+the regular Run() function.
+
+Also, there is the RunHelper class that simplifies working with the Initialize-Step-Run-Step-Final
+cycle writing it as a single expression (see an example).
+
+Four more functions helpful in lexical analysis are LongestPrefix(), LongestSuffix(), ShortestPrefix(),
+and ShortestSuffix(). These functions returns the largest / the smallest prefix or suffix allowed by
```

Comment:
It is better to say that the FSM "matches" or "accepts" the given text, rather than "allows".